### PR TITLE
Fix reading from MergeTree with non-deterministic functions in filter

### LINKED ISSUE #61106: query with rand condition gives non-expected results (maybe related to virtual projection)
https://fiddle.clickhouse.com/755a4164-6fc1-43af-9ea4-f840e23b2ecd (rerun that several times)

```
CREATE TABLE users (uid Int16, age Int16) ENGINE=MergeTree order by uid;

INSERT INTO users select number, rand()%10 from numbers(1000000);

select '-- count';
SELECT count() FROM users;

select '-- rand()%2=0:';
SELECT count() FROM users WHERE rand()%2=0;
```

/cc @amosbird

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix unexpected result during reading from tables with virtual columns when filter contains non-deterministic functions. Closes https://github.com/ClickHouse/ClickHouse/issues/61106.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 7632c2c33f357c1c616f734c7bf2502ccbfbd496 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60586/7632c2c33f357c1c616f734c7bf2502ccbfbd496/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/VirtualColumnUtils.cpp [hunk 1]
```diff
@@ -467,6 +467,23 @@ static bool canEvaluateSubtree(const ActionsDAG::Node * node, const Block & allo
     return true;
 }
 
+static bool isDeterministicInScopeOfQuery(const ActionsDAG::Node * node)
+{
+    if (node->type != ActionsDAG::ActionType::FUNCTION)
```
Comment (KochetovNicolai) [inline src/Storages/VirtualColumnUtils.cpp:472]: One possible node type is ALIAS, so let's maybe check all children first.

**Review by KochetovNicolai — APPROVED** [review]
Ok, but let's fix aliases for `isDeterministicInScopeOfQuery`

**Comment by clickgapai** [comment]
Hi @Avogar @KochetovNicolai — while reviewing this PR I found the following:

- **Coverage gap:** Trivial count with AND of partition column and non-deterministic function on partitioned `MergeTree` is untested — https://github.com/ClickHouse/ClickHouse/pull/103893

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**